hash.c

```diff
@@ -880,10 +880,11 @@ ar_general_foreach(VALUE hash, st_foreach_check_callback_func *func, st_update_c
                 return 0;
               case ST_REPLACE:
                 if (replace) {
+                    ar_table_pair *orig_pair = pair;
                     (*replace)(&key, &val, arg, TRUE);
 
-                    // TODO: pair should be same as pair before.
                     pair = RHASH_AR_TABLE_REF(hash, i);
+                    HASH_ASSERT(orig_pair == pair);
```

Comment:
Reading this, I think what the original TODO may have intended is that in non-debug builds, maybe we don't need to do the `RHASH_AR_TABLE_REF` lookup a second time? What do you think?

```suggestion
                    (*replace)(&key, &val, arg, TRUE);

                    // Pair should not have moved
                    HASH_ASSERT(pair == RHASH_AR_TABLE_REF(hash, i));
```

Reply:
This makes sense to me. I tried committing your patch but got the error:

> Applying suggestions on deleted lines is currently not supported.

I'm going to manually apply your patch and push it for review.

I'd love for @ko1 to comment on his original intent for this TODO.a